Expose handshake_timeout option in listeners

rbino opened the pull request:
## Proposed Changes

Ranch had a default value of 5 seconds for the handshake timeout. This sometimes caused constrained devices to fail the connection since they couldn't complete the SSL handshake in that amount of time.

This commit allows configuring the handshake timeout to arbitrary values to support such use cases.

This also includes a small internal refactoring, renaming the old `transport_opts` to `socket_opts`, since that's what they actually are, and introducing new base `transport_opts` that depend on the ranch transport.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes issue #1670)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CODE_OF_CONDUCT.md` document
- [ ] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if needed)
- [x] Any dependent changes have been merged and published in related repositories
- [x] I have updated changelog (At the bottom of the release version)
- [x] I have squashed all my commits into one before merging

ioolkos commented:
@rbino Thanks.
- Can you take out the documentation for TCP listeners in the schema file (unless, of course, you see a use case.. but as you mentioned to go into no-op anyway)

Just FYI: We're currently fixing the tests, and some fail on Github actions anyway.

rbino commented:
@ioolkos should I just delete the documentation or should I delete everything (i.e. keep `handshake_timeout` only on SSL transports like `mqtts`, `wss`, `https`...)?

ioolkos commented:
@rbino apologies, missed your questions above. I'd keep the `handshake_timeout` option only for SSL transports, yes.

rbino commented:
@ioolkos updated. I also updated the merge description, in particular I added:

This also includes a small internal refactoring, renaming the old transport_opts to socket_opts, since that's what they actually are, and introducing new base transport_opts that depend on the ranch transport.

ioolkos commented:
@rbino do you want to keep this Draft?

rbino commented:
@ioolkos marked as Ready for Review. I tried running tests locally but I've found some unrelated failures - are these the same which are failing in the CI?

rbino commented:
@ioolkos btw I ran a VerneMQ image built from my branch in a full Astarte system against the hardware that was failing and I verified this actually works and fixes the problem when increasing the timeout to 30 seconds using environment variables

ratonlaveur38 commented:
@rbino  , I am planning to give it a try with our device in China => usually fails 5s timeout.

After compilation, I was checking the vernemq,conf generated by `make rel` and could not find the new parameters. `listener.https.tls_handshake_timeout` etc ... 

Not sure how this file is generated and might be that I missed something (in this case discard the comment  :-) )

ioolkos commented:
@ratonlaveur38 those options are added as `hidden` configs so as to not clutter the conf file. (and most users will not set them)
You can just add them to the conf file yourself when you need to set them.
